packages/@dcos/sorting/index.ts

```diff
@@ -0,0 +1,67 @@
+export enum ComparisonResult {
+  GREATER_THAN = 1,
+  EQUAL = 0,
+  LESS_THAN = -1
+}
+export type Comparator<T> = (a: T, b: T) => ComparisonResult;
+export type Getter<I, T> = (data: I) => T;
+type compareWithGetter<T> = <I>(getter: Getter<I, T>) => Comparator<I>;
+
+export function sort<T>(comparators: Array<Comparator<T>>, data: T[]): T[] {
+  return [...data].sort((a, b) => {
+    for (const comparator of comparators) {
```

Comment:
not blocking, but please consider thinking about for a second: 
I'd love to get rid of this for loop:

```
[...data].sort((a,b) => sortByComparators(a, b, comparators))
```

```
function sortByComparators<T>(a: T, b: T, comparators: Array<Comparator<T>>): ComparisonResult {
  if(comparators.length === 0) return 0;
  return (comparators[0](a, b) === 0) ? 0 : sortByComparators(a, b, comparators.slice(1));
}
```

Reply:
Why do you want to get rid of this for loop?

Reply:
Loops are error prone, adding complexity and are hard to reason about - even the simple ones. Also being able to think of a loop as recursive function calls makes it a lot easier to write complete, secure and generally "good" code. 

there are two general best practices violated in this particular loop:
1. never break a loop (instead write your condition so that it stops at the correct time)
2. never have if clauses (*) inside loops (again, write your loop condition appropriately)

* early exists might be okay in some cases (i know that this team likes them - personally I dont)

Reply:
I honestly think that in this specific case adding a new function and having a recursion to traverse a list makes it harder to reason about the code. 

The for loop is pretty clear, because the if is an early exit. 

But I don't care enough, I'll just rewrite it, whatever 🤷‍♂️

Reply:
> The for loop is pretty clear, because the if is an early exit.

well, this is the reason why I dont like early exits.
```
function this_is_ok(bar) {
  if(!bar) return 1;
}
function this_is_not(bar) {
  const baz = do_weird_shit_with(bar);
  if(!baz) return 1;
}
```

> But I don't care enough, I'll just rewrite it, whatever 🤷‍♂️
Awesome…

Reply:
I still see no problem with this example, but as I said, going to change it.